src/accountsdb/db.zig

```diff
@@ -0,0 +1,1645 @@
+
+const Logger = @import("../trace/log.zig").Logger;
+const Level = @import("../trace/level.zig").Level;
+
+const printTimeEstimate = @import("../time/estimate.zig").printTimeEstimate;
+
+const _accounts_index = @import("index.zig");
+const AccountIndex = _accounts_index.AccountIndex;
+const DiskMemoryConfig = _accounts_index.DiskMemoryConfig;
+const RamMemoryConfig = _accounts_index.RamMemoryConfig;
+const RefMemoryLinkedList = _accounts_index.RefMemoryLinkedList;
+const AccountRef = _accounts_index.AccountRef;
+const AccountIndexBin = _accounts_index.AccountIndexBin;
+
+pub const MERKLE_FANOUT: usize = 16;
+pub const ACCOUNT_INDEX_BINS: usize = 8192;
+
+pub const AccountsDBConfig = struct {
+    // number of Accounts to preallocate for cache
+    storage_cache_size: usize = 10_000,
+    // number of bins to shard the index pubkeys across -- must be power of two
+    n_index_bins: usize = ACCOUNT_INDEX_BINS,
+    // how many RAM references to preallocate for each bin
+    index_ram_capacity: usize = 0,
+    // where to create disk indexes files (if null, will not use disk indexes)
+    disk_index_dir: ?[]const u8 = null,
+    // how many DISK references to preallocate for each bin
+    index_disk_capacity: usize = 0,
+};
+
+/// database for accounts
+pub const AccountsDB = struct {
+    allocator: std.mem.Allocator,
+
+    // holds the account data
+    storage: AccountStorage,
+    // maps a pubkey to the account location
+    index: AccountIndex,
+
+    logger: Logger,
+    config: AccountsDBConfig,
+    fields: AccountsDbFields = undefined,
+
+    const Self = @This();
+
+    pub fn init(
+        allocator: std.mem.Allocator,
+        logger: Logger,
+        config: AccountsDBConfig,
+    ) !Self {
+        const storage = try AccountStorage.init(
+            allocator,
+            config.storage_cache_size,
+        );
+
+        var disk_index_config: ?DiskMemoryConfig = null;
+        if (config.disk_index_dir != null) {
+            disk_index_config = DiskMemoryConfig{
+                .dir_path = config.disk_index_dir.?,
+                .capacity = config.index_disk_capacity,
+            };
+        }
+
+        const index = try AccountIndex.init(
+            allocator,
+            config.n_index_bins,
+            RamMemoryConfig{
+                .capacity = config.index_ram_capacity,
+            },
+            disk_index_config,
+        );
+
+        return Self{
+            .storage = storage,
+            .index = index,
+            .allocator = allocator,
+            .logger = logger,
+            .config = config,
+        };
+    }
+
+    pub fn deinit(self: *Self) void {
+        self.storage.deinit();
+        self.index.deinit(true);
+        self.logger.deinit();
+    }
+
+    /// used to build AccountsDB from a snapshot in parallel
+    pub const LoadingThreadAccountsDB = struct {
+        index: AccountIndex,
+        file_map: std.AutoArrayHashMap(FileId, AccountFile),
```

Comment:
talked about this offline - but decided it would be safer to stick to `std` unless performance is an issue since not all std methods are implemented on SwissMap